### PR TITLE
[metadata-update] Make a copy of the DIL when applying updates

### PR DESCRIPTION
**Pull request by lambdageek**
Otherwise we may point into memory that is cleaned up by the GC

### CODE REVIEW

**Comment by lambdageek** [comment]
Would like to backport this to 6.0-preview4

**Review by thaystg — APPROVED** [review]
LGTM

File: src/mono/mono/metadata/metadata-update.c [hunk 1]
```diff
@@ -423,6 +423,17 @@ mono_image_open_dmeta_from_data (MonoImage *base_image, uint32_t generation, gco
 	return dmeta_image;
 }
 
+static gpointer
+open_dil_data (MonoImage *base_image G_GNUC_UNUSED, gconstpointer dil_src, uint32_t dil_length)
+{
+	/* TODO: find a better memory manager.  But this way we at least won't lose the IL data. */
+	MonoMemoryManager *mem_manager = (MonoMemoryManager *)mono_alc_get_default ()->memory_manager;
```
Comment (lambdageek) [inline src/mono/mono/metadata/metadata-update.c:430]: I know I could use `m_image_get_memory_manager (base_image)` here, but that function looks different in 6.0-preview4, so for the sake of the backport, I'd rather dump the delta IL into the default memory manager for now.

**Comment by lambdageek** [comment]
Verified via manual testing that we no longer crash if dil_data is corrupted (or ends up pointing at reclaimed memory) by changing

https://github.com/dotnet/runtime/blob/a56a142b65b322f276281f039032848bc8e51ac0/src/mono/sample/mbr/DeltaHelper/DeltaHelper.cs#L62

to 

```csharp
            LoadMetadataUpdate (assm, dmeta_data, dil_data, dpdb_data);
            for (int i = 0; i < dil_data.Length; ++i)
                    dil_data[i] = (byte)255;
```

**Pull request merged**